tracer/src/Datadog.Trace/Ci/Agent/CIWriterHttpSender.cs

```diff
@@ -0,0 +1,171 @@
+                }
+                catch (Exception ex)
+                {
+                    Log.Error(ex, "An error occurred while generating http request to send events to {AgentEndpoint}", _apiRequestFactory.Info(tracesEndpoint));
+                    return;
+                }
+
+                bool success = false;
+                Exception exception = null;
+                bool isFinalTry = retryCount >= retryLimit;
+
+                try
+                {
+                    success = await SendPayloadAsync(new ArraySegment<byte>(payloadBytes), payloadMimeType, numberOfTraces, request, isFinalTry).ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    exception = ex;
+#if DEBUG
+                    if (ex.InnerException is InvalidOperationException ioe)
+                    {
+                        Log.Error<int, string>(ex, "An error occurred while sending {Count} events to {AgentEndpoint}", numberOfTraces, _apiRequestFactory.Info(tracesEndpoint));
+                        return;
+                    }
+#endif
+                }
+
+                // Error handling block
+                if (!success)
+                {
+                    if (isFinalTry)
+                    {
+                        // stop retrying
+                        Log.Error<int, string>(exception, "An error occurred while sending {Count} events to {AgentEndpoint}", numberOfTraces, _apiRequestFactory.Info(tracesEndpoint));
+                        return;
+                    }
+
+                    // Before retry delay
+                    bool isSocketException = false;
+                    Exception innerException = exception;
+
+                    while (innerException != null)
+                    {
+                        if (innerException is SocketException)
+                        {
+                            isSocketException = true;
+                            break;
+                        }
+
+                        innerException = innerException.InnerException;
+                    }
+
+                    if (isSocketException)
+                    {
+                        Log.Debug(exception, "Unable to communicate with {AgentEndpoint}", _apiRequestFactory.Info(tracesEndpoint));
+                    }
+
+                    // Execute retry delay
+                    await Task.Delay(sleepDuration).ConfigureAwait(false);
+                    retryCount++;
+                    sleepDuration *= 2;
+
+                    continue;
+                }
+
+                Log.Debug<int, string>("Successfully sent {Count} events to {AgentEndpoint}", numberOfTraces, _apiRequestFactory.Info(tracesEndpoint));
+                return;
+            }
+        }
+
+        private async Task<bool> SendPayloadAsync(ArraySegment<byte> payload, string mimeType, int numberOfTraces, IApiRequest request, bool finalTry)
```

Comment:
Looks like you can get rid of the `int numberOfTraces` parameter

```suggestion
        private async Task<bool> SendPayloadAsync(ArraySegment<byte> payload, string mimeType, IApiRequest request, bool finalTry)
```

Reply:
done in [4f6d0d7](https://github.com/DataDog/dd-trace-dotnet/pull/2402/commits/4f6d0d7a9dbede063444bb55459def841342f20a)